### DISCUSSION TITLE
[@types/prismjs] Grammar type is not usable

### DISCUSSION THREAD

**Discussion by srmagura** [DefinitelyTyped/DefinitelyTyped#58208] [Issues with a @types package]
The code

```ts
const myGrammar: Prism.Grammar = {
  'my-sub-grammar': {
    pattern: /arbitraryRegex/
  },
  rest: Prism.languages.jsx
}
```

produces the TS error

```text
Type '{ 'my-sub-grammar': { pattern: RegExp; }; rest: Prism.Grammar; }' is not assignable to type 'Grammar'.
  Type '{ 'my-sub-grammar': { pattern: RegExp; }; rest: Prism.Grammar; }' is not assignable to type 'Record<string, GrammarValue>'.
    Property 'rest' is incompatible with index signature.
      Type 'Grammar' is not assignable to type 'GrammarValue'.
        Type 'Grammar' is missing the following properties from type '(RegExp | TokenObject)[]': length, pop, push, concat, and 28 more.ts(2322)
```

I'm trying to figure out if this is an issue with `@types/prismjs` or TypeScript itself.  https://github.com/microsoft/TypeScript/issues/47456

[a comment by an automated account is not shown]

**Comment by MysteryBlokHed** [comment]
It looks like the reason this error is happening is because of the `Grammar` type defined here in @types/prismjs:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/64952b7a8c790d4a7b1085cdc650fcd86a3f0d3d/types/prismjs/index.d.ts#L172

Changing this line to the following fixed it:

```typescript
export type Grammar = GrammarRest | Record<string, GrammarValue>; 
```

I'm not sure why this fix actually works, since it seems like the current definition should be fine.

**Reply by srmagura** [reply]
That fixes it! (Though, it is weird to me that TypeScript unions work that way.) PR submitted: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58387

**Comment by srmagura** [comment]
PR submitted: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/58387
